smartsim/_core/_cli/verify.py

```diff
@@ -0,0 +1,189 @@
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import argparse
+import io
+import tempfile
+import typing as t
+from contextlib import contextmanager
+from types import TracebackType
+
+import numpy as np
+from smartredis import Client
+
+from smartsim import Experiment
+from smartsim._core.utils.helpers import installed_redisai_backends
+from smartsim._core._cli.utils import SMART_LOGGER_FORMAT
+from smartsim.log import get_logger
+
+SMART_LOGGER_FORMAT = "[%(name)s] %(levelname)s %(message)s"
+logger = get_logger("Smart", fmt=SMART_LOGGER_FORMAT)
+
+# Many of the functions in this module will import optional
+# ML python packages only if they are needed to verify the build is working
+#
+# pylint: disable=import-error,import-outside-toplevel
+# mypy: disable-error-code="import"
+
+
+class _VerificationTempDir(tempfile.TemporaryDirectory):
+    """A Temporary directory to be used as a context manager that will only
+    clean itself up if no error is raised within its context
+    """
+
+    def __exit__(
+        self,
+        exc: t.Optional[t.Type[BaseException]],
+        value: t.Optional[BaseException],
+        tb: t.Optional[TracebackType],
+    ) -> None:
+        if not value:  # Yay, no error! Clean up as normal
+            super().__exit__(exc, value, tb)
+        else:  # Uh-oh! Better make sure this is not implicitly cleaned up
+            self._finalizer.detach()  # type: ignore[attr-defined]
+
+
+def execute(_args: argparse.Namespace, /) -> int:
+    backends = installed_redisai_backends()
+    try:
+        with _VerificationTempDir() as temp_dir:
+            verify_install(
+                location=temp_dir,
+                port=8734,  # TODO: allow users to pass as arg `--port`?
+                with_tf="tensorflow" in backends,
+                with_pt="torch" in backends,
+                with_onnx="onnxruntime" in backends,
+            )
+    except Exception as e:
+        logger.error(
+            "SmartSim failed to run a simple experiment!\n"
+            f"Experiment failed due to the following exception:\n{e}\n\n"
+            f"Output files are available at `{temp_dir}`"
+        )
+        return 2
+    return 0
+
+
+def verify_install(
+    location: str, port: int, with_tf: bool, with_pt: bool, with_onnx: bool
+) -> None:
+    exp = Experiment("Verification", exp_path=location, launcher="local")
+    with _make_managed_orc(exp, port) as client:
```

Comment:
Since this verify check is always going to happen on the local. Can we add a check to make sure that the port is actually open. I'm a little bit worried about hitting false negatives just because it can't find a port. Throwing a more informative error that at least says, 'Eh things might work, but we can't verify because the port is being used'. 

```
import socket

with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as redis_socket:
  if redis_socket.connect_ex(('localhost',port)) ~= 0:
    raise SSConnectionError(f'Port {port} cannot be bound. Cannot verify RedisAI installation.')
  redis_socket.close()
```
     



If you wanted to be really fancy, give a list of a few non-commonly used ports and iterate through them to find one that's actually open.